Dockerfile

```diff
@@ -1,112 +1,160 @@
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+    jq \
+    && rm -rf /var/lib/apt/lists/*
+
+# Define packages to install
+ENV PACKAGES="torch torchvision torchaudio"
+
+# Step 1: Preinstall Dependencies (Cached Separately)
+RUN set -eux; \
+    if echo "${INSTALL_CHANNEL}" | grep -q "nightly"; then \
+    PIP_PRE_FLAG="--pre"; \
+    else \
+    PIP_PRE_FLAG=""; \
+    fi; \
+    if [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
+    EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu/"; \
+    else \
+    EXTRA_INDEX_URL="https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH%.}/"; \
+    fi; \
+    echo "Using EXTRA_INDEX_URL=${EXTRA_INDEX_URL} with PIP_PRE_FLAG=${PIP_PRE_FLAG}"; \
+    for PACKAGE in $PACKAGES; do \
+    echo "Fetching dependencies for $PACKAGE"; \
+    curl -s https://pypi.org/pypi/${PACKAGE}/json -o ${PACKAGE}.json; \
+    DEPS=$(jq -r '.info.requires_dist[]? | split(";")[0]' ${PACKAGE}.json | tr '\n' ' '); \
+    if [ -n "$DEPS" ]; then \
+    echo "Installing dependencies for $PACKAGE: $DEPS"; \
+    pip install --no-cache-dir $DEPS --extra-index-url $EXTRA_INDEX_URL; \
+    else \
+    echo "No dependencies found for $PACKAGE"; \
+    fi; \
+    rm -f ${PACKAGE}.json; \
+    done
+
+# Step 2: Install Main Torch Packages Without Dependencies
+RUN set -eux; \
+    if echo "${INSTALL_CHANNEL}" | grep -q "nightly"; then \
+    PIP_PRE_FLAG="--pre"; \
+    else \
+    PIP_PRE_FLAG=""; \
+    fi; \
+    if [ "$TARGETPLATFORM" = "linux/arm64" ]; then \
+    EXTRA_INDEX_URL="https://download.pytorch.org/whl/cpu/"; \
+    else \
+    EXTRA_INDEX_URL="https://download.pytorch.org/${INSTALL_CHANNEL}/${CUDA_PATH%.}/"; \
+    fi; \
+    echo "Installing main packages with PIP_PRE_FLAG=${PIP_PRE_FLAG}"; \
+    pip install --no-cache-dir $PACKAGES --no-deps $PIP_PRE_FLAG --extra-index-url $EXTRA_INDEX_URL && \
+    pip install torchelastic && \
     /opt/conda/bin/conda clean -ya
-RUN /opt/conda/bin/pip install torchelastic
-RUN IS_CUDA=$(python -c 'import torch ; print(torch.cuda._is_compiled())'); \
-    echo "Is torch compiled with cuda: ${IS_CUDA}"; \
-    if test "${IS_CUDA}" != "True" -a ! -z "${CUDA_VERSION}"; then \
-        exit 1; \
-    fi
 
+# Stage 6: Official Environment Setup
 FROM ${BASE_IMAGE} as official
 ARG PYTORCH_VERSION
 ARG TRITON_VERSION
 ARG TARGETPLATFORM
 ARG CUDA_VERSION
 LABEL com.nvidia.volumes.needed="nvidia_driver"
 RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
-        ca-certificates \
-        libjpeg-dev \
-        libpng-dev \
-        && rm -rf /var/lib/apt/lists/*
-COPY --from=conda-installs /opt/conda /opt/conda
-RUN if test -n "${TRITON_VERSION}" -a "${TARGETPLATFORM}" != "linux/arm64"; then \
-        DEBIAN_FRONTEND=noninteractive apt install -y --no-install-recommends gcc; \
-        rm -rf /var/lib/apt/lists/*; \
+    ca-certificates \
+    libjpeg-dev \
+    libpng-dev \
```

Comment:
Let me know if we want to add `g++` here in the case we want also to close https://github.com/pytorch/pytorch/issues/116696

Reply:
`g++` is going to solve the use `torch.compile` in runtime images as we have discussed in https://github.com/pytorch/pytorch/issues/116696 (fixes: #116696). I've maintaned the original `libjpeg-dev` and `libpng-dev` but I don't know why there are currently in the runtime images.